pandas/core/interval.py

```diff
@@ -0,0 +1,413 @@
+        if (locs == -1).any():
+            raise KeyError("a list-indexer must only include existing intervals")
+
+        return locs
+
+    def _check_method(self, method):
+        if method is not None:
+            raise NotImplementedError(
+                'method %r not yet implemented for IntervalIndex' % method)
+
+    def _searchsorted_monotonic(self, label, side, exclude_label=False):
+        if not self.is_non_overlapping_monotonic:
+            raise KeyError('can only get slices from an IntervalIndex if '
+                           'bounds are non-overlapping and all monotonic '
+                           'increasing or decreasing')
+
+        if isinstance(label, IntervalMixin):
+            raise NotImplementedError
+
+        if ((side == 'left' and self.left.is_monotonic_increasing) or
+                (side == 'right' and self.left.is_monotonic_decreasing)):
+            sub_idx = self.right
+            if self.open_right or exclude_label:
+                label = _get_next_label(label)
+        else:
+            sub_idx = self.left
+            if self.open_left or exclude_label:
+                label = _get_prev_label(label)
+
+        return sub_idx._searchsorted_monotonic(label, side)
+
+    def _get_loc_only_exact_matches(self, key):
+        return self._multiindex._tuple_index.get_loc(key)
+
+    def _find_non_overlapping_monotonic_bounds(self, key):
+        if isinstance(key, IntervalMixin):
+            start = self._searchsorted_monotonic(
+                key.left, 'left', exclude_label=key.open_left)
+            stop = self._searchsorted_monotonic(
+                key.right, 'right', exclude_label=key.open_right)
+        else:
+            # scalar
+            start = self._searchsorted_monotonic(key, 'left')
+            stop = self._searchsorted_monotonic(key, 'right')
+        return start, stop
+
+    def get_loc(self, key, method=None):
+        self._check_method(method)
+
+        original_key = key
+
+        if self.is_non_overlapping_monotonic:
+            if isinstance(key, Interval):
+                left = self._maybe_cast_slice_bound(key.left, 'left', None)
+                right = self._maybe_cast_slice_bound(key.right, 'right', None)
+                key = Interval(left, right, key.closed)
+            else:
+                key = self._maybe_cast_slice_bound(key, 'left', None)
+
+            start, stop = self._find_non_overlapping_monotonic_bounds(key)
+
+            if start + 1 == stop:
+                return start
+            elif start < stop:
+                return slice(start, stop)
+            else:
+                raise KeyError(original_key)
+
+        else:
+            # use the interval tree
+            if isinstance(key, Interval):
+                left, right = _get_interval_closed_bounds(key)
+                return self._tree.get_loc_interval(left, right)
+            else:
+                return self._tree.get_loc(key)
+
+    def get_value(self, series, key):
```

Comment:
@jreback this method on existing Index objects is a complete mess, but unfortunately it's called by `series[key]` so it's pretty necessary. Any ideas how to handle it in a sane, minimal way?